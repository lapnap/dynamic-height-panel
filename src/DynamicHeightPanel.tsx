// Libraries
import React, {PureComponent, CSSProperties} from 'react';

// Types
import {Options} from './types';
import {PanelProps} from '@grafana/ui';

export interface Props extends PanelProps<Options> {}

interface State {
  height: number;
  time: number;
  up: boolean;
  running: boolean;
}

export class DynamicHeightPanel extends PureComponent<Props, State> {
  frameId: number = 0;

  constructor(props: Props) {
    super(props);

    this.state = {
      height: props.options.min,
      time: Date.now(),
      up: true,
      running: false,
    };
  }

  componentDidMount() {
    this.startLoop();
  }

  componentWillUnmount() {
    this.stopLoop();
  }

  startLoop = () => {
    if (this.frameId == 0) {
      this.frameId = window.requestAnimationFrame(this.loop);
      this.setState({
        running: true,
      });
    }
  };

  stopLoop = () => {
    window.cancelAnimationFrame(this.frameId);
    this.frameId = 0;
    this.setState({
      running: true,
    });
  };

  loop = () => {
    const {height, time, up, running} = this.state;
    const {min, max, speed} = this.props.options;
    const now = Date.now();
    const elapsed = (now - time) / 1000; // in seconds
    const dist = elapsed * speed * (up ? 1 : -1);

    if (!running) {
      return;
    }

    let dir = up;
    let next = height + dist;
    if (next > max) {
      next = max;
      dir = !dir;
    } else if (next < min) {
      next = min;
      dir = !dir;
    }
    this.setState({
      height: next,
      up: dir,
      time: now,
    });

    // Set up next iteration of the loop\
    this.frameId = window.requestAnimationFrame(this.loop);
  };

  toggleRunning = () => {
    if (this.frameId) {
      this.stopLoop();
    } else {
      this.startLoop();
    }
  };

  render() {
    const {height, running} = this.state;

    const style: CSSProperties = {
      height,
      border: this.props.options.border,
      padding: '10px',
    };

    return (
      <div style={style}>
        <button className="btn btn-secondary" onClick={this.toggleRunning}>
          {running ? 'Stop' : 'Start'}
        </button>
        height: {height}
      </div>
    );
  }
}
