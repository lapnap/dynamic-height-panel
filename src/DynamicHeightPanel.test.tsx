import React from 'react';
import {shallow} from 'enzyme';
import {DynamicHeightPanel, Props} from './DynamicHeightPanel';

const setup = (propOverrides?: object) => {
  const props: Props = {
    height: 300,
    width: 300,
  } as Props; // partial

  Object.assign(props, propOverrides);

  const wrapper = shallow(<DynamicHeightPanel {...props} />);
  const instance = wrapper.instance() as DynamicHeightPanel;

  return {
    instance,
    wrapper,
  };
};

describe('Render Presense Panel with basic options', () => {
  it('should render', () => {
    const {wrapper} = setup();
    expect(wrapper).toBeDefined();
    expect(wrapper).toMatchSnapshot();
  });
});
