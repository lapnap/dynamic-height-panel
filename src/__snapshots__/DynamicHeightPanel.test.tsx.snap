// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Panel with basic options should render 1`] = `
<div
  style={
    Object {
      "border": undefined,
      "height": 100,
      "padding": "10px",
    }
  }
>
  <button
    className="btn btn-secondary"
    onClick={[Function]}
  >
    Stop
  </button>
  height: 
  100
</div>
`;
